homeassistant/components/rpi_gpio/binary_sensor.py

```diff
@@ -64,15 +80,35 @@ def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
 
         def read_gpio(port):
             """Read state from GPIO."""
-            self._state = rpi_gpio.read_input(self._port)
+            if self._read:
+                self._state = rpi_gpio.read_input(self._port)
+            else:
+                self._state = not self._invert_logic
             self.schedule_update_ha_state()
 
-        rpi_gpio.edge_detect(self._port, read_gpio, self._bouncetime)
+        # edge detection mode:
+        # - edge: use GPIO state after any edge without polling
+        # - poll: use GPIO state after any edge with polling
+        # - sticky: only detect rising/falling edge and use polling to reset
+        if edge_mode is EdgeModes.edge:
+            self._poll = False
+            self._edge = "BOTH"
+            self._read = True
+        elif edge_mode is EdgeModes.poll:
+            self._poll = True
+            self._edge = "BOTH"
+            self._read = True
+        elif edge_mode is EdgeModes.sticky:
+            self._poll = True
+            self._edge = "FALLING" if self._invert_logic else "RISING"
+            self._read = False
+
+        rpi_gpio.edge_detect(self._port, self._edge, read_gpio, self._bouncetime)
```

Comment:
do that on `async def async_added_to_hass(self)` as executor job

Reply:
@pvizeli thank you for looking at this PR 👏👍

Reply:
The callback should be first enable after the entity is added to the state machine, otherwise, it runs into errors if they receive a callback before.

Reply:
@pvizeli Got it. Is it ok to have `read_gpio` be a local function in `async_added_to_hass`? Also does `rpi_gpio.edge_detect` need to be called async or something?

Reply:
@ntessore can I help in any way? Maybe I can help with tests (if You tell me how to modify those files in HA). I'm pulling my hair for the last couple of days. I have test environment ready :)

Reply:
yeah async_added_to_hass is async and you need use `await hass.async_run_executor_job(my_funct, param1, param2)`